From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shyam Jeedigunta <jeedigv@amazon.com>
Date: Mon, 19 Jul 2021 12:31:37 -0700
Subject: [PATCH] --EKS-PATCH-- Ignore 'wait: no child processes' error when
 calling mount/umount

Cherry-pick of upstream Kubernetes:
https://github.com/kubernetes/kubernetes/pull/103780, which was
included in Kubernetes 1.23.

This change fixes an issue related to a race condition. See this issue:
https://github.com/kubernetes/kubernetes/issues/103753

From the original PR description:
    I've only fixed the exec commands that are part of Mount() and
    Unmount() functions and that too in the linux mount helper. Not
    touching others, since I'm not sure about the implications.

Signed-off-by: Kirsten Schumy <ksschumy@amazon.com>
---
 staging/src/k8s.io/mount-utils/mount_linux.go | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/staging/src/k8s.io/mount-utils/mount_linux.go b/staging/src/k8s.io/mount-utils/mount_linux.go
index 9e6b033cb4b..031e9f0e1df 100644
--- a/staging/src/k8s.io/mount-utils/mount_linux.go
+++ b/staging/src/k8s.io/mount-utils/mount_linux.go
@@ -43,6 +43,8 @@ const (
 	fsckErrorsCorrected = 1
 	// 'fsck' found errors but exited without correcting them
 	fsckErrorsUncorrected = 4
+	// Error thrown by exec cmd.Run() when process spawned by cmd.Start() completes before cmd.Wait() is called (see - k/k issue #103753)
+	errNoChildProcesses = "wait: no child processes"
 )
 
 // Mounter provides the default implementation of mount.Interface
@@ -177,6 +179,14 @@ func (mounter *Mounter) doMount(mounterPath string, mountCmd string, source stri
 	command := exec.Command(mountCmd, mountArgs...)
 	output, err := command.CombinedOutput()
 	if err != nil {
+		if err.Error() == errNoChildProcesses {
+			if command.ProcessState.Success() {
+				// We don't consider errNoChildProcesses an error if the process itself succeeded (see - k/k issue #103753).
+				return nil
+			}
+			// Rewrite err with the actual exit error of the process.
+			err = &exec.ExitError{ProcessState: command.ProcessState}
+		}
 		klog.Errorf("Mount failed: %v\nMounting command: %s\nMounting arguments: %s\nOutput: %s\n", err, mountCmd, mountArgsLogStr, string(output))
 		return fmt.Errorf("mount failed: %v\nMounting command: %s\nMounting arguments: %s\nOutput: %s",
 			err, mountCmd, mountArgsLogStr, string(output))
@@ -280,6 +290,14 @@ func (mounter *Mounter) Unmount(target string) error {
 	command := exec.Command("umount", target)
 	output, err := command.CombinedOutput()
 	if err != nil {
+		if err.Error() == errNoChildProcesses {
+			if command.ProcessState.Success() {
+				// We don't consider errNoChildProcesses an error if the process itself succeeded (see - k/k issue #103753).
+				return nil
+			}
+			// Rewrite err with the actual exit error of the process.
+			err = &exec.ExitError{ProcessState: command.ProcessState}
+		}
 		return fmt.Errorf("unmount failed: %v\nUnmounting arguments: %s\nOutput: %s", err, target, string(output))
 	}
 	return nil
